core/kernel/user_ta.c

```diff
@@ -477,15 +477,18 @@ TEE_Result tee_ta_init_user_ta_session(const TEE_UUID *uuid,
 
 	utc->ta_ctx.ts_ctx.uuid = *uuid;
 	res = vm_info_init(&utc->uctx, &utc->ta_ctx.ts_ctx);
-	if (res)
-		goto out;
+	if (res) {
+		condvar_destroy(&utc->ta_ctx.busy_cv);
+		free_utc(utc);
+		return res;
+	}
+
 	utc->ta_ctx.is_initializing = true;
 
 #ifdef CFG_TA_PAUTH
 	crypto_rng_read(&utc->uctx.keys, sizeof(utc->uctx.keys));
 #endif
 
-	mutex_lock(&tee_ta_mutex);
```

Comment:
Would adding an `assert(mutex_is_locked(&tee_ta_mutex));` make sense here or at the beginning of this function?
We don't have a `mutex_is_locked()` function, but that's trivial to implement.

Reply:
`mutex_try_lock()` would be better because `mutex_is_locked()` cannot be reliable when it returns false.

Reply:
`mutex_is_locked()` would never be reliable, but it would make the intent clear and catch enough errors that we'd understand something is wrong.
`mutex_try_lock()` here would in the best case be deceiving since we don't know which thread is holding the lock.

Reply:
With `mutex_is_locked()` you would not know which thread is holding the lock either. Perhaps we need `mutex_is_locked_by_current_thread()`.

Reply:
That requires further bookkeeping. The point with the assert isn't a 100% catch rate. If it works while there's no further contention on the mutex that's good enough. Another option is to only leave a comment here explaining that we expect the mutex to be held, but that will quite likely have an even lower catch rate. ;-)

Reply:
If you think it's safer in the long term to keep the mutex lock/unlock calls near where it's needed (from inside each pta/uta/stmm context/session setup functions), we can discard this session init functions split and come back to the previous patch series proposal.

At the time,  your concern was that `tee_ta_init_session_with_context()` was called twice 
(https://github.com/OP-TEE/optee_os/pull/6836#discussion_r1615578427). I think we can remove the 1st call, from _tee_ta_manager.c_. _pseudo_ta.c_ would call it once, only if UUID matches a PTA UUID. _stmm_sp.c_ would call it only if  the UUID matches StMM interface one. _user_ta.c_ would call it once.

Reply:
No, I'm not concerned. A comment is fine as a reminder if there's a problem with an assert.

Reply:
I've appended a fixup commit to address this discussion thread. Please tell me if it matches your expectations.